Add culprit extraction for captured exceptions

kanadaj opened the pull request:
Addresses https://github.com/elastic/apm-agent-dotnet/issues/705 and https://github.com/elastic/apm-agent-dotnet/issues/145. Iterates through the exception's StackFrames to find the first frame with the module name not in the exclusion blacklist.

Might be worth adding an inclusion list (to override blacklist items) or a whitelist too (only include namespaces on the whitelist).

kanadaj commented:
CLA signed now

kanadaj commented:
@gregkalapos any feedback on this one?

Also, why is the config handling so incredibly over-complicated...

gregkalapos commented on src/Elastic.Apm/Model/ExecutionSegmentCommon.cs:
```diff
@@ -164,6 +165,46 @@ internal static Task<T> CaptureSpan<T>(Span span, Func<ISpan, Task<T>> func)
+			foreach (var frame in frames)
+			{
+				var method = frame.GetMethod();
+				var module = method.DeclaringType?.FullName ?? "Unknown Module";
```
nit: Wouldn't be `fullyQualifiedTypeName` a better name for this variable?  - since that's what it is, it's not really the module. Also I'd rename the parameter name of `IsInApp`.

kanadaj commented on src/Elastic.Apm/Model/ExecutionSegmentCommon.cs:
```diff
@@ -164,6 +165,46 @@ internal static Task<T> CaptureSpan<T>(Span span, Func<ISpan, Task<T>> func)
+				var method = frame.GetMethod();
+				var module = method.DeclaringType?.FullName ?? "Unknown Module";
+				if (IsInApp(module,excludedModules))
+				{
```
Will apply these and do a new commit

kanadaj commented:
@gregkalapos 

I feel like the whitelist based approach from Java may not suit all use cases, especially for larger projects with lots of 1st party libraries as they might need way too many inclusions to be useful. It would also mean ElasticApm wouldn't have suitable defaults for error reporting, since setting a default value for the inclusion list is a tall order.

I feel the best approach then would be to have both an inclusion list (like Java) and an exclusion list. If the inclusion list is defined in the config, we would use that, otherwise fall back to the exclusion list which would be also configurable.

gregkalapos commented:
> I feel the best approach then would be to have both an inclusion list (like Java) and an exclusion list. If the inclusion list is defined in the config, we would use that, otherwise fall back to the exclusion list which should be also configurable.

Yes, I agree with this. The [python agent](https://www.elastic.co/guide/en/apm/agent/python/current/configuration.html#config-include-paths) has something similar.

Only last thing to decide is the naming of these 2 configs.

Python calls them `include_paths` and `exclude-paths` - I like that the name suggest that those are related - but using the names `paths` would not fit .NET in my opinion.


Some ideas:
- `IncludedNamespaces`,  `ExcludedNamespaces` (alternatively we could also use `include` instead of `included` similar to python. 
- `ApplicationNamespaces`, `FrameworkNamespaces` (or `LibraryNamespaces`)
- Or we mix those, so keep `ExcludedNamespaces` and add `ApplicationNamespaces`

I prefer `ApplicationNamespaces`, and  `FrameworkNamespaces`. Happy to take opinion.

Also, the PR is fine with just 1 setting - the next one can be in another PR or I can also take care of that - also we should add docs for this, which is also fine in another PR.

Currently working on a patch release, once that's out I'll merge things to `master`.

kanadaj commented:
The mixed naming feels the most right, since excluded namespaces might include namespaces not part of the framework, e.g. logging frameworks, other crash reporting, 3rd party middlewares, etc. I'll do a push tonight with the added setting, it shouldn't take all that much work to do.

kanadaj commented:
Added. I've also added some more unit tests for this. It doesn't feel like `BasicAgentTests` is quite the right place for this, but I wasn't quite sure where to put them, so they probably should be moved.

gregkalapos commented:
This breaks some of our API tests, plus we also have an overload where users can manually set the `culprit`, I think that is also broken. I'll take care of that part.

gregkalapos commented:
Ah, the overload that sets the culprit manually works fine, those tests are still green - only `Elastic.Apm.Tests.ApiTests.ApiTests.ErrorOnTransaction` fails. Adapted that in #740.
